atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/HostnamesByIpSupplier.java

```diff
@@ -43,20 +45,27 @@ public final class HostnamesByIpSupplier implements Supplier<Map<String, String>
     private static final String HOSTNAME_COLUMN = "hostname";
     private static final String IP_COLUMN = "ip";
 
-    private final Supplier<PoolingContainer<CassandraClient>> randomGoodHostSupplier;
+    private final Supplier<List<PoolingContainer<CassandraClient>>> hosts;
 
-    public HostnamesByIpSupplier(Supplier<PoolingContainer<CassandraClient>> randomGoodHostSupplier) {
-        this.randomGoodHostSupplier = randomGoodHostSupplier;
+    public HostnamesByIpSupplier(Supplier<List<PoolingContainer<CassandraClient>>> hosts) {
+        this.hosts = hosts;
     }
 
     @Override
     public Map<String, String> get() {
-        try {
-            return randomGoodHostSupplier.get().runWithPooledResource(getHostnamesByIp());
-        } catch (Exception e) {
-            log.warn("Could not get hostnames by ip from Cassandra", e);
-            return ImmutableMap.of();
-        }
+        return hosts.get().stream()
+                .map(container -> {
+                    try {
+                        return Optional.ofNullable(container.runWithPooledResource(getHostnamesByIp()));
+                    } catch (Exception e) {
+                        log.warn("Could not get hostnames by ip from Cassandra", e);
+                        return Optional.<Map<String, String>>empty();
+                    }
+                })
+                .filter(Optional::isPresent)
+                .map(Optional::get)
+                .findFirst()
+                .orElseGet(ImmutableMap::of);
```

Comment:
makes sense - though for the timer, this might still take a while if the runWithPooledResource blocks (though maybe we have a bound of that, based on the socket / connect timeouts)